ciao-controller/api/api.go

```diff
@@ -0,0 +1,516 @@
+	}
+
+	err = c.MapAddress(req.PoolName, req.InstanceID)
+	if err != nil {
+		return errorResponse(err), err
+	}
+
+	return Response{http.StatusNoContent, nil}, nil
+}
+
+func unmapExternalIP(c *Context, w http.ResponseWriter, r *http.Request) (Response, error) {
+	vars := mux.Vars(r)
+	tenantID, ok := vars["tenant"]
+	mappingID := vars["mapping_id"]
+
+	var IPs []types.MappedIP
+
+	if !ok {
+		IPs = c.ListMappedAddresses(nil)
+	} else {
+		IPs = c.ListMappedAddresses(&tenantID)
+	}
+
+	for _, m := range IPs {
+		if m.ID == mappingID {
+			err := c.UnMapAddress(m.ExternalIP)
+			if err != nil {
+				return errorResponse(err), err
+			}
+
+			return Response{http.StatusAccepted, nil}, nil
+		}
+	}
+
+	return errorResponse(types.ErrAddressNotFound), types.ErrAddressNotFound
+}
+
+// Service is an interface which must be implemented by the ciao API context.
+type Service interface {
+	AddPool(name string, subnet *string, ips []string) (types.Pool, error)
+	ListPools() ([]types.Pool, error)
+	ShowPool(id string) (types.Pool, error)
+	DeletePool(id string) error
+	AddAddress(poolID string, subnet *string, IPs []string) error
+	RemoveAddress(poolID string, subnetID *string, IPID *string) error
+	ListMappedAddresses(tenantID *string) []types.MappedIP
+	MapAddress(poolName *string, instanceID string) error
+	UnMapAddress(ID string) error
+}
+
+// Context is used to provide the services and current URL to the handlers.
+type Context struct {
+	URL string
+	Service
+}
+
+// Config is used to setup the Context for the ciao API.
+type Config struct {
+	URL         string
+	CiaoService Service
+}
+
+// Routes returns the supported ciao API endpoints.
+// A plain application/json request will return v1 of the resource
+// since we only have one version of this api so far, that means
+// most routes will match both json as well as our custom
+// content type.
+func Routes(config Config) *mux.Router {
+	// make new Context
+	context := &Context{config.URL, config.CiaoService}
+
+	r := mux.NewRouter()
+
+	// external IP pools
+	route := r.Handle("/", Handler{context, listResources})
+	route.Methods("GET")
+
+	route = r.Handle("/{tenant:[a-f0-9]{8}-?[a-f0-9]{4}-?4[a-f0-9]{3}-?[8|9|aA|bB][a-f0-9]{3}-?[a-f0-9]{12}}", Handler{context, listResources})
+	route.Methods("GET")
+
+	matchContent := fmt.Sprintf("application/(%s|json)", PoolsV1)
+
+	route = r.Handle("/pools", Handler{context, listPools})
+	route.Methods("GET")
+	route.HeadersRegexp("Content-Type", matchContent)
+
+	route = r.Handle("/{tenant:[a-f0-9]{8}-?[a-f0-9]{4}-?4[a-f0-9]{3}-?[8|9|aA|bB][a-f0-9]{3}-?[a-f0-9]{12}}/pools", Handler{context, listPools})
+	route.Methods("GET")
+	route.HeadersRegexp("Content-Type", matchContent)
+
+	route = r.Handle("/pools", Handler{context, addPool})
+	route.Methods("POST")
+	route.HeadersRegexp("Content-Type", matchContent)
+
+	route = r.Handle("/pools/{pool:[a-f0-9]{8}-?[a-f0-9]{4}-?4[a-f0-9]{3}-?[8|9|aA|bB][a-f0-9]{3}-?[a-f0-9]{12}}", Handler{context, showPool})
```

Comment:
@rbradford it matches the ciao-cli commands we used. Since we have versioning now, we can
change if we need to in the future, and IMO it's nice to have the resource name short as possible.